### PR TITLE
chore(#8433) add telemetry events to contact forms

### PR DESCRIPTION
**Pull request by Benmuiruri**
# Description

- Adds[ Telemetry Events ](https://docs.communityhealthtoolkit.org/apps/guides/performance/telemetry/) to record Telemetry data in the contact forms 

### Personal Targets

- Each "card" is separated into two content areas
- The content areas is shaded differently to delineate them
- The labels and images which used to be on the top, are now at the lower content area
- Instead of using red for unmet targets, the cards now use black

#8433  

# Code review checklist

- [ ] Readable: Concise, well named, follows the [style guide](https://docs.communityhealthtoolkit.org/contribute/code/style-guide/), documented if necessary.
- [ ] Documented: Configuration and user documentation on [cht-docs](https://github.com/medic/cht-docs/)
- [ ] Tested: Unit and/or e2e where appropriate
- [ ] Internationalised: All user facing text
- [ ] Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in the release notes.

# Compose URLs

If Build CI hasn't passed, these may 404:

- [Core](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:template-issue-update/docker-compose/cht-core.yml)
- [CouchDB Single](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:template-issue-update/docker-compose/cht-couchdb.yml)
- [CouchDB Cluster](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:template-issue-update/docker-compose/cht-couchdb-clustered.yml)

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

### CODE REVIEW

File: webapp/src/ts/modules/contacts/contacts-edit.component.ts [hunk 1]
```diff
@@ -29,6 +30,7 @@ export class ContactsEditComponent implements OnInit, OnDestroy, AfterViewInit {
     private contactTypesService:ContactTypesService,
     private dbService:DbService,
     private contactSaveService:ContactSaveService,
+    private telemetryService: TelemetryService,
```
Comment (garethbowen) [inline webapp/src/ts/modules/contacts/contacts-edit.component.ts:33]: Nit: whitespace consistency.

```suggestion
    private telemetryService:TelemetryService,
```

**Review by garethbowen — COMMENTED** [review]
Very nice @Benmuiruri . Thanks for remembering to update the unit tests too.

I've put a very minor suggestion in there.

Before I approve it though I had two questions I'd like your take on...

1. Did you think about end-to-end testing this to make sure that the whole stack was working together? Do you think unit testing is sufficient for this change?
2. I see a lot of copied code from the reports component. Did you consider creating a new service that both components could call to make the code a little simpler? What about moving the code down into the enketo service? I'm not even sure that would be possible but curious if you explored it...

**Review by tatilepizs — COMMENTED** [review]
We have been talking, in our QA syncs, about when it is necessary to include e2e test and when we can add only unit tests or integration tests. And we have talked that maybe it will be a good idea to establish a "rule" that every time that we make changes to a file in the webapp folder we should add an e2e new test or a scenario to an existing one.  (we need to put this in a document, we are working on it 😅 )

What do you think?

We can definitely help you in the process, let me know and we can have a call about it 🙂

**Comment by Benmuiruri** [comment]
Hi @garethbowen, 

1. I didn't consider adding e2e tests given how it had been done in the other components, but here are my thoughts about it.

- Since we are interested in ensuring that the data is not only generated but also sent to and correctly handled by the backend, then e2e would be good to add. 
- A unit test may not be sufficient because it does not how multiple components and external services interact, it only allows us to test that we are recording the data. 
- In the unit tests we are mocking several external interactions, arguably that might not cover all possible conditions and edge cases that we can't see in unit tests. 

2. On the issue of duplication, I did realise it was not the DRYest code. It's definitely a good idea to encapsulate the shared logic in a service and inject it to any component that reuses it.  Perhaps think of `Component Inheritance`, where we have a `Base Component`  that other components can extend.... Maybe...

However, I'll be honest and admit at this time, I do not fully understand `CHT`, `Angular` and `Enketo` enough at this time to evaluate the coupling and cohesion and be able to refactor the implementation and the tests. 

I'd certainly welcome the challenge to improve the code health by making the code DRY at a later time when I have a better understanding of the code.

**Comment by Benmuiruri** [comment]
Hi @tatilepizs,  I would love to have a chat about recommended thinking and approaches to testing `cht`. Please set it up, my calendar is flexible 😃

**Review by garethbowen — APPROVED** [review]
@Benmuiruri Thanks for the thoughtful response!

Approved as I'm happy with where this is at now, but feel free to continue to improve testing or DRYness as you see fit!
